[Alerting]: Implement test rule API route

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
If the recipient is grafana evaluates the condition. Otherwise, it proxies to ~`/query_range`~`/query` route of the respective data source.

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:
Requires grafana/alerting-api#46

papagian committed:
[Alerting]: Implement test rule API route

papagian committed:
Merge remote-tracking branch 'origin/master' into alerting-test-rule-api

papagian committed:
missing modifications

papagian committed:
Update to latest alerting-api

papagian committed:
Update to latest alerting-api

papagian committed:
lint

owen-d commented on pkg/services/ngalert/api/api_testing.go:
```diff
@@ -0,0 +1,85 @@
+
+		switch ds.Type {
+		case "loki":
+			path = "loki/api/v1/query_range"
```
Rules are evaluated at one particular moment in time (now) and actually would use the `/query` instead of `/query_range` under the hood. I can see it being helpful, however, to have the history that `/query_range` provides in a test endpoint. Is this what you're intending to do?

papagian replied:
1. I had noticed the both `/query` and `/query_range` are called but I thought that for our case they were interchangeable so I just picked one. So no, it was not intentional.
2. We could pass a specific time as a query parameter (same as it happens for the grafana rules) so that we could test the query for different times. WDYT?

owen-d commented on pkg/services/ngalert/api/api_testing.go:
```diff
@@ -0,0 +1,85 @@
+	params.Set("start", strconv.FormatInt(start.Unix(), 10))
+	params.Set("end", strconv.FormatInt(end.Unix(), 10))
+	params.Set("directtion", "BACKWARD")
+	params.Set("limit", strconv.Itoa(1332))
```
Why 1332 here? I don't think the limit param should be set at all here as it's only valid for Loki and only for fetching a limited number of logs.

owen-d replied:
```suggestion
```

owen-d commented on pkg/services/ngalert/api/api_testing.go:
```diff
@@ -0,0 +1,85 @@
+	params.Set("query", body.Expr)
+	params.Set("start", strconv.FormatInt(start.Unix(), 10))
+	params.Set("end", strconv.FormatInt(end.Unix(), 10))
+	params.Set("directtion", "BACKWARD")
```
```suggestion
```

owen-d commented on pkg/services/ngalert/api/api_testing.go:
```diff
@@ -0,0 +1,85 @@
+	params.Set("end", strconv.FormatInt(end.Unix(), 10))
+	params.Set("directtion", "BACKWARD")
+	params.Set("limit", strconv.Itoa(1332))
+	params.Set("step", strconv.Itoa(15))
```
Step should probably not be hardcoded. It has the potential to be inefficient or throw errors if there are too many steps, based on the time range of the query. I'd suggest using something like (plucked from Loki):
```go
// defaultQueryRangeStep returns the default step used in the query range API,
// which is dynamically calculated based on the time range
func defaultQueryRangeStep(start time.Time, end time.Time) int {
	return int(math.Max(math.Floor(end.Sub(start).Seconds()/250), 1))
}
```

owen-d replied:
We should actually remove the `step` parameter as it's used only for `/query_range`.

owen-d reviewed (commented):
I have some questions/suggestions about the integrations

papagian committed:
Apply suggestions from code review

papagian committed:
Merge remote-tracking branch 'origin/master' into alerting-test-rule-api

papagian committed:
Call /query instead of /query_range

papagian committed:
Fix failing API test

With the latest alerting-api, there is a failing API
test (due to unexpected payload).
This commit downgrades the alerting-api
until the failing test issue will be resolved.

papagian committed:
Merge remote-tracking branch 'origin/master' into alerting-test-rule-api

papagian committed:
Merge remote-tracking branch 'origin/master' into alerting-test-rule-api

papagian committed:
Merge remote-tracking branch 'origin/master' into alerting-test-rule-api

papagian committed:
remove accidentally commited files
